File: pl_stress/rules.py
"""
Rule-based Polish stress resolver.

Rules encoded from:
  - akcentowanie.md (Wikipedia article + Kutnyj + Polszczyzna.pl + academic sources)
  - Rada Języka Polskiego guidelines

Polish stress is PENULTIMATE by default. Rules below handle exceptions.

Return value convention throughout this module:
  stress_from_end: int  — 1-based, counting from word end
    1 = last syllable (oxytone)
    2 = penultimate (DEFAULT — rules only return this explicitly for clitic compounds)
    3 = antepenultimate (proparoksytona)
    4 = pre-antepenultimate
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional

from .syllabify import count_syllables


class Confidence(Enum):
    EXACT = auto()   # found in exception dictionary
    RULE = auto()    # matched a productive grammatical rule
    DEFAULT = auto() # penultimate fallback


@dataclass(frozen=True)
class RuleMatch:
    stress_from_end: int
    rule_name: str


# ---------------------------------------------------------------------------
# Rule predicates
# ---------------------------------------------------------------------------

def _is_acronym(word: str) -> bool:
    """
    All-uppercase ASCII letter strings are acronyms → oxytone (stress on last syllable).
    E.g.: PKP, ONZ, KGB, TVP, GPS
    Mixed-case words with digits/special chars are excluded.
    """
    return len(word) >= 2 and all(ch.isupper() and ch.isascii() for ch in word)


# Prefixes that produce oxytone when attached to a monosyllabic noun:
_OXYTONE_PREFIXES = ("arcy", "eks", "wice", "mini", "super", "ekstra", "euro")


def _is_prefix_plus_monosyllable(word: str) -> bool:
    """
    Words like arcymistrz, eksmąż, wicekról, minibar → oxytone.
    Heuristic: known prefix + the rest has exactly ONE vowel.
    """
    lower = word.lower()
    for prefix in _OXYTONE_PREFIXES:
        if lower.startswith(prefix) and len(lower) > len(prefix):
            rest = lower[len(prefix):]
            if count_syllables(rest) == 1:
                return True
    return False


# ---------------------------------------------------------------------------
# Past tense plural (1st/2nd person) → antepenultimate (3rd from end)
# Endings: -liśmy, -łyśmy, -liście, -łyście
# Source: all references, consistently
# ---------------------------------------------------------------------------
_PAST_PLURAL_ENDINGS = ("liśmy", "łyśmy", "liście", "łyście")


def _is_past_plural(word: str) -> bool:
    lower = word.lower()
    return any(lower.endswith(e) for e in _PAST_PLURAL_ENDINGS)


# ---------------------------------------------------------------------------
# Conditional mood singular + 3rd plural → antepenultimate
# Endings: -łbym, -łbyś, -łby, -łaby, -łabyś, -łabym, -łoby,
#          -liby, -łyby, -łoby, -nęłby, etc.
# Pattern: contains -by- or ends with -by after a consonant cluster
# ---------------------------------------------------------------------------
_COND_SG_ENDINGS = (
    "łbym", "łbyś", "łby",
    "łabym", "łabyś", "łaby",
    "łobym", "łobyś", "łoby",
    "liby", "łyby",
    # rarer forms
    "nęłbym", "nęłbyś", "nęłby",
    "słbym", "słbyś", "słby",
)


def _is_conditional_singular(word: str) -> bool:
    lower = word.lower()
    return any(lower.endswith(e) for e in _COND_SG_ENDINGS)


# ---------------------------------------------------------------------------
# Conditional mood plural (1st/2nd) → pre-antepenultimate (4th from end)
# Endings: -libyśmy, -libyście, -łybyśmy, -łybyście, -łobyśmy, -łobyście
# ---------------------------------------------------------------------------
_COND_PL_ENDINGS = (
    "libyśmy", "libyście",
    "łybyśmy", "łybyście",
    "łobyśmy", "łobyście",
    "słibyśmy", "słibyście",
)


def _is_conditional_plural(word: str) -> bool:
    lower = word.lower()
    return any(lower.endswith(e) for e in _COND_PL_ENDINGS)


# ---------------------------------------------------------------------------
# Nouns ending in -ika/-yka (Latin loans) → antepenultimate
# But only when nominative singular OR case with SAME syllable count.
# We can't know the case here, so we apply broadly and note as Rule
# (PoliMorf propagation will correct specific forms).
#
# Also applies to person-nouns in genitive singular with same syllable count:
# cybernetykiem (same syllable count as cybernetyk)
# ---------------------------------------------------------------------------
def _is_ika_yka(word: str) -> bool:
    lower = word.lower()
    return lower.endswith(("ika", "yka")) and count_syllables(lower) >= 3


# ---------------------------------------------------------------------------
# Numerals: 400–900 with -sta/-set/-kroć → antepenultimate
# Specific words, hard-coded (the productive pattern is very limited)
# ---------------------------------------------------------------------------
_NUMERAL_ANTEPENULT: frozenset[str] = frozenset({
    "czterysta", "czterystu",
    "pięćset", "pięciuset", "pięciuset",
    "sześćset", "sześciuset",
    "siedemset", "siedmiuset",
    "osiemset", "ośmiuset",
    "dziewięćset", "dziewięciuset",
    "osiemkroć", "siedemkroć", "sześćkroć", "pięciokroć",
    "czterokroć", "tysiąckroć",
})
# czterystoma is PENULTIMATE (exception to the exception)
_NUMERAL_PENULT_EXCEPTION: frozenset[str] = frozenset({"czterystoma"})


def _is_numeral_antepenult(word: str) -> bool:
    lower = word.lower()
    return lower in _NUMERAL_ANTEPENULT and lower not in _NUMERAL_PENULT_EXCEPTION


# ---------------------------------------------------------------------------
# Conjunctions with personal suffixes / conditional morphemes → antepenultimate
# abyśmy, żebyście, jeśliby, ponieważby, etc.
# ---------------------------------------------------------------------------
_CONJ_ROOTS = ("aby", "żeby", "ażeby", "jeśliby", "jeżeliby",
               "iżby", "ponieważby", "gdyby")
_CONJ_ANTEPENULT_ENDINGS = ("śmy", "ście", "by", "bym", "byś", "byście", "byśmy")


def _is_conjunction_antepenult(word: str) -> bool:
    lower = word.lower()
    # Exact compound check
    for root in _CONJ_ROOTS:
        if lower.startswith(root) and len(lower) > len(root):
            suffix = lower[len(root):]
            if any(suffix == e or suffix.endswith(e) for e in _CONJ_ANTEPENULT_ENDINGS):
                return count_syllables(lower) >= 3
    return False


# ---------------------------------------------------------------------------
# Traditional foreign words with antepenultimate stress (normative list)
# ---------------------------------------------------------------------------
_TRADITIONAL_ANTEPENULT: frozenset[str] = frozenset({
    "komitet", "uniwersytet", "prezydent",
    "maksimum", "minimum", "optimum",
    "muzeum", "liceum", "technikum",
    "audytorium", "laboratorium", "gymnasium", "seminarium",
    "centrum", "collegium",
})


def _is_traditional_antepenult(word: str) -> bool:
    return word.lower() in _TRADITIONAL_ANTEPENULT


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def apply_rules(word: str) -> Optional[RuleMatch]:
    """
    Apply productive stress rules to a word.
    Returns RuleMatch or None (caller should fall back to penultimate default).

    Rules are checked in priority order (most specific first).
    """
    lower = word.lower()
    n = count_syllables(lower)

    # 1. Acronyms → oxytone (last syllable)
    if _is_acronym(word):  # use original case
        return RuleMatch(stress_from_end=1, rule_name="acronym")

    # 2. Prefix + monosyllable → oxytone
    if _is_prefix_plus_monosyllable(lower):
        return RuleMatch(stress_from_end=1, rule_name="prefix_monosyllable")

    if n < 3:
        # All remaining rules require ≥ 3 syllables
        return None

    # 3. Conditional plural (1st/2nd person) → 4th from end
    if n >= 4 and _is_conditional_plural(lower):
        return RuleMatch(stress_from_end=4, rule_name="conditional_plural")

    # 4. Past plural (1st/2nd person) → antepenultimate
    if _is_past_plural(lower):
        return RuleMatch(stress_from_end=3, rule_name="past_plural")

    # 5. Conditional singular + 3rd plural → antepenultimate
    if _is_conditional_singular(lower):
        return RuleMatch(stress_from_end=3, rule_name="conditional_singular")

    # 6. Conjunctions with conditional/personal morphemes → antepenultimate
    if _is_conjunction_antepenult(lower):
        return RuleMatch(stress_from_end=3, rule_name="conjunction")

    # 7. Numerals 400–900 → antepenultimate
    if _is_numeral_antepenult(lower):
        return RuleMatch(stress_from_end=3, rule_name="numeral")

    # 8. Latin-loan -ika/-yka nouns → antepenultimate
    if _is_ika_yka(lower):
        return RuleMatch(stress_from_end=3, rule_name="ika_yka")

    # 9. Traditional foreign words → antepenultimate
    if _is_traditional_antepenult(lower):
        return RuleMatch(stress_from_end=3, rule_name="traditional_foreign")

    return None

File: crates/core/src/rules.rs
use crate::syllabify::count_syllables;
use crate::Confidence;

/// Productive grammatical rules for Polish stress exceptions.
///
/// Returns `Some((syllable_index, confidence))` if the word matches a rule,
/// where `syllable_index` is 0-based from the start and `n` is the total
/// syllable count (pre-computed by the caller from `syllabify`).
pub fn apply_rules(word: &str, n: usize) -> Option<(usize, Confidence)> {
    let lower = word.to_lowercase();

    // --- Acronyms / abbreviations: oxytone (last syllable) ---
    // Heuristic: all characters are ASCII uppercase letters.
    if word.len() >= 2 && word.chars().all(|c| c.is_ascii_uppercase()) {
        return Some((n.saturating_sub(1), Confidence::Rule));
    }

    // --- One-syllable stems with arcy-/eks-/wice- prefix: oxytone ---
    let oxytone_prefixes = ["arcy", "eks", "wice"];
    for prefix in &oxytone_prefixes {
        if let Some(rest) = lower.strip_prefix(prefix) {
            if count_syllables(rest) == 1 {
                return Some((n.saturating_sub(1), Confidence::Rule));
            }
        }
    }

    if n < 3 {
        return None; // rules below only apply to 3+ syllable words
    }

    // --- 1st/2nd person plural past tense: antepenultimate (3rd from end) ---
    // Endings: -liśmy, -łyśmy, -liście, -łyście
    let past_plural = ["liśmy", "łyśmy", "liście", "łyście"];
    for ending in &past_plural {
        if lower.ends_with(ending) {
            return Some((n - 3, Confidence::Rule));
        }
    }

    // --- Conditional mood, singular + 3rd plural: antepenultimate ---
    // Endings: -łbym, -łbyś, -łby, -łaby, -łabyś, -łabym, -liby, -łyby
    let cond_singular = [
        "łbym", "łbyś", "łby", "łaby", "łabyś", "łabym",
        "liby", "łyby", "łoby",
    ];
    for ending in &cond_singular {
        if lower.ends_with(ending) {
            return Some((n - 3, Confidence::Rule));
        }
    }

    // --- Conditional mood, 1st/2nd plural: pre-antepenultimate (4th from end) ---
    // Endings: -libyśmy, -libyście, -łybyśmy, -łybyście
    if n >= 4 {
        let cond_plural = ["libyśmy", "libyście", "łybyśmy", "łybyście"];
        for ending in &cond_plural {
            if lower.ends_with(ending) {
                return Some((n - 4, Confidence::Rule));
            }
        }
    }

    // --- Words ending in -ika/-yka (Latin loans): antepenultimate ---
    if lower.ends_with("ika") || lower.ends_with("yka") {
        return Some((n - 3, Confidence::Rule));
    }

    // --- Conjunctions with conditional/personal morphemes: antepenultimate ---
    // abyśmy, żebyście, jeśliby, etc.
    let conj_prefixes = ["aby", "żeby", "ażeby", "jeśli", "jeżeli", "iżby", "ponieważ"];
    for prefix in &conj_prefixes {
        if lower.starts_with(prefix) && lower.len() > prefix.len() {
            let rest = &lower[prefix.len()..];
            if matches!(rest, "śmy" | "ście" | "by" | "bym" | "byś") {
                return Some((n - 3, Confidence::Rule));
            }
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_acronym_oxytone() {
        // PKP: 1 syllable → index 0 (last = only)
        let n = crate::count_syllables("PKP");
        assert_eq!(apply_rules("PKP", n), Some((n.saturating_sub(1), Confidence::Rule)));
    }

    #[test]
    fn test_past_plural_antepenultimate() {
        // zrobiliśmy: 5 syllables → index 2
        let n = crate::count_syllables("zrobiliśmy");
        let r = apply_rules("zrobiliśmy", n);
        assert!(r.is_some());
        let (idx, conf) = r.unwrap();
        assert_eq!(conf, Confidence::Rule);
        assert_eq!(idx, n - 3);
    }

    #[test]
    fn test_conditional_plural_preantepenultimate() {
        // zrobilibyśmy: 6 syllables → index 2
        let n = crate::count_syllables("zrobilibyśmy");
        let r = apply_rules("zrobilibyśmy", n);
        assert!(r.is_some());
        let (idx, _) = r.unwrap();
        assert_eq!(idx, n - 4);
    }

    #[test]
    fn test_ika_yka_antepenultimate() {
        let n = crate::count_syllables("fizyka");
        assert_eq!(apply_rules("fizyka", n), Some((n - 3, Confidence::Rule)));
        let n = crate::count_syllables("matematyka");
        assert_eq!(apply_rules("matematyka", n), Some((n - 3, Confidence::Rule)));
    }

    #[test]
    fn test_no_match_default_word() {
        let n = crate::count_syllables("kobieta");
        assert_eq!(apply_rules("kobieta", n), None);
    }
}
